lib/aiken-design-patterns/singular-utxo-indexer-one-to-many.ak

```diff
@@ -1,44 +1,105 @@
-use aiken/list
-use aiken/transaction.{Input, Output, ScriptContext, Spend, Transaction}
+use aiken/collection/list
+use aiken_design_patterns/utils
+use cardano/transaction.{
+  Input, NoDatum, Output, OutputReference, Transaction, placeholder,
+}
 
+/// Spend endpoint helper function. By including this in your validator, you'll
+/// an efficient access to your input, its corresponding outputs, total output
+/// value and count.
+///
+/// > [!NOTE]
+/// > This function has no protection against
+/// > [double satisfaction](https://github.com/Plutonomicon/plutonomicon/blob/b6906173c3f98fb5d7b40fd206f9d6fe14d0b03b/vulnerabilities.md#double-satisfaction)
+/// > vulnerability, as this can be done in multiple ways depending on the
+/// > contract. If you can tolerate some extra overhead, consider using the
```

Comment:
I do not think that there is any protocol that might not want to avoid a double satisfaction attack. Maybe we could implement a countermeasure for it in this module?

Reply:
I can think of 4 ways to prevent double satisfaction:
1. Equality of input index and output index (not applicable everywhere)
2. Datum tagging (would require an additional function for extracting the tag field from the output datum)
3. Filtering the inputs and expecting one to remain (most versatile, but costly)
4. Expecting the continuing output at index 0, such that it holds only 1 NFT (works for contracts that only hold multiple beacon UTxOs)

Picking one over the others might make the interface somewhat opinionated. What do you think?